rfcs/text/0015_ols_phase_1.md

```diff
@@ -0,0 +1,260 @@
+
+There are two primary changes to this wrapper:
+
+#### Attaching ACLs
+
+This wrapper will be responsible for attaching an ACL to all private objects before they are created.
+It will also allow users to provide their own ACL in order to support the import/export use cases.
+
+Similar to the way we treat `namespaces`, it will not be possible to change an ACL via the `update`/`bulk_update` functions in this first phase. We may consider adding a dedicated function to update the ACL, similar to what we've done for sharing to spaces.
+
+#### Authorization changes
+
+This wrapper will be updated to ensure that access to private objects is only granted to authorized users. A user is authorized to operate on a private saved object if **all of the following** are true:
+Step 1) The user is authorized to perform the operation on saved objects of the requested type, within the requested space. (Example: `update` a `user-settings` saved object in the `marketing` space)
+Step 2) The user is authorized to access this specific instance of the saved object, as described by that object's ACL. For this first phase, the `acl.owner` is allowed to perform all operations. The only other users who are allowed to access this object are administrators (see [unresolved question 2](#82-authorization-for-private-objects))
+
+Step 1 of this authorization check is the same check we perform today for all existing saved object types. Step 2 is a new authorization check, and **introduces additional overhead and complexity**. We explore the logic for this step in more detail later in this RFC.
+
+![High-level authorization model for private objects](../images/ols_phase_1_auth.png)
+
+## 3.2 Saved Objects API
+
+OLS Phase 1 does not introduce any new APIs, but rather augments the existing Saved Object APIs.
+
+APIs which return saved objects are augmented to include the top-level `acl` property when it exists. This includes the `export` API.
+
+APIs that create saved objects are augmented to accept an `acl` property. This includes the `import` API.
+
+### `get` / `bulk_get`
+
+The security wrapper will ensure the user is authorized to access private objects before returning them to the consumer.
+
+#### Performance considerations
+None. The retrieved object contains all of the necessary information to authorize the current user, with no additional round trips to Elasticsearch.
+
+### `create` / `bulk_create`
+
+The security wrapper will ensure that an ACL is attached to all private objects.
+
+If the caller has requested to overwrite existing `private` objects, then the security wrapper must ensure that the user is authorized to do so.
+
+#### Performance considerations
+When overwriting existing objects, the security wrapper must first retrieve all of the existing `private` objects to ensure that the user is authorized. This requires another round-trip to `get`/`bulk-get` all `private` objects so we can authorize the operation.
```

Comment:
Thanks for calling this out, I meant to explain my thinking here. I agree that an additional pre-flight request from the security wrapper is less than ideal, but I'm hesitant to over-engineer an optimal solution in this first phase, since I don't _expect_ that we'll make heavy use of private multi-space objects in the near future.

Refresh my memory: why do the current set of pre-flight namespace checks live in the repository? I want to make sure I'm not overlooking a scenario where the repository should also be validating the `accessControl`.

Reply:
When you take an action on an object, the repository has to retrieve it and make sure the object exists in the current space context. This was not needed when we had single-namespace objects (whose raw document IDs contained the namespace), but now that we have multi-namespace objects we need to do that preflight check.

At the moment the preflight check only occurs on multi-namespace object types, which is rare, but in the future most object types will be multi-namespace...

Reply:
Sorry I should have been clearer: why does the _repository_ have to perform the check, as opposed to say, the spaces wrapper?

Reply:
Because if the Spaces plugin is disabled, the repository should still behave this way (e.g., you should only be able to interact with objects that were in the default space before the Spaces plugin was disabled)

Reply:
Would it increase the complexity to take advantage of the existing pre-flight checks? It seems like it would simplify it because these pre-flights are already being done. The security wrapper could just ensure that the specified owner matches the currently authenticated end-user, and then pass the owner to the repository that would then enforce this logic.

Reply:
Agree with @kobelb. We decided to build the namespace infrastructure into the repository and the same might make sense for private objects. Especially now that licensing is not the primary driver for how we architect these layers.

Reply:
> Would it increase the complexity to take advantage of the existing pre-flight checks? It seems like it would simplify it because these pre-flights are already being done. The security wrapper could just ensure that the specified owner matches the currently authenticated end-user, and then pass the owner to the repository that would then enforce this logic.

It would complicate audit logging a bit. Currently, we perform audit logging in the wrapper before we attempt to actually create/update/delete the requested object. Having security perform its own pre-flight check allows us to audit the operation before the repository attempts to do any work. If we take advantage of the repository's pre-flight check, then we'll have a few options for how we decide to audit:

1) Leave the audit logging as-is. This means that the audit log will not record events where the repository has rejected requests because of a mismatched `owner`.
2) Change audit logging so that we wait for a response from the repository before recording the event. This leads to an increased risk of missed events if the operation was successful, but we failed to get a response back due to a networking glitch or software error.
3) Allow the repository to signal when it's "safe" to audit the event before sending the request to ES. There are a few ways we could accomplish this (event-based, callback, introduce a core audit logger...), and we'd need to think through the best approach to serve our needs going forward. I'm not opposed to exploring these options, but this is the type of work I was hoping to avoid coupling to this RFC.

Reply:
It seems like this auditing limitation might come up in a couple of other places. Do we have the same problem when:
 - updating a `namespaceType: 'multiple'` saved objects that exists, but not in the current namespace?
 - getting or updating a saved object that doesn't exist?
 - performing an operation which doesn't succeed because of a network error?

> I'm not opposed to exploring these options, but this is the type of work I was hoping to avoid coupling to this RFC.

I'm quite nervous about the complexity of saved objects as a whole. My impression is that there's only a handful of people that somewhat grasp how it all fits together (and I find myself constantly learning some new behaviour like audit logging limitations). This implementation complexity also comes with a very complex set of tests where the amount of permutations keep steadily growing. So anything that could lower complexity feels worthwhile to explore.

We can always decide to adopt some technical debt / complexity in the short term, but I think a large part of the value of an RFC is starting to shape a more concrete long term technical vision so that if we adopt additional complexity we have some fleshed out ideas for how we can pay it down.

Reply:
> It seems like this auditing limitation might come up in a couple of other places. Do we have the same problem when:
> 
> * updating a `namespaceType: 'multiple'` saved objects that exists, but not in the current namespace?
> * getting or updating a saved object that doesn't exist?
> * performing an operation which doesn't succeed because of a network error?

Yes, these situations are all possible today. We "mitigate" this by explaining that the audit log [records an outcome of `unknown`](https://www.elastic.co/guide/en/kibana/current/xpack-security-audit-logging.html#_category_database) for these scenarios. If you want a definitive answer, then you have to correlate Kibana audit logs with the corresponding Elasticsearch audit logs to be sure that the request actually completed. We can definitively tell you when the request has failed due to authorization checks in the wrapper, but beyond that the audit log will effectively tell the auditor that the user was authorized to perform this action, but we don't know if it completed successfully or not.



> I'm quite nervous about the complexity of saved objects as a whole. My impression is that there's only a handful of people that somewhat grasp how it all fits together (and I find myself constantly learning some new behaviour like audit logging limitations). This implementation complexity also comes with a very complex set of tests where the amount of permutations keep steadily growing. So anything that could lower complexity feels worthwhile to explore.

No disagreement here on any of this!



> We can always decide to adopt some technical debt / complexity in the short term, but I think a large part of the value of an RFC is starting to shape a more concrete long term technical vision so that if we adopt additional complexity we have some fleshed out ideas for how we can pay it down.

I can spend some time on this if that'd be helpful. Long-term, it seems like the structure of the saved objects service may change somewhat drastically (i.e. potentially deprecating the SOC and generic APIs). I know we've had some long-form discussions, but do we have anything more concise that explains what we're targeting? We probably do, I just can't find it 🙂.
If I know what we're targeting, then I can try to tailor the approach to match that vision.
If not, I can enumerate these other ideas with the assumption that the current architecture will remain largely in place, and then we can figure out what that looks like in the evolved architecture.

Reply:
> do we have anything more concise that explains what we're targeting

We need another RFC 🙈 

I'm working on writing down a longer term strategy in 7.13. I think the most impactful change for auditing might be deprecating the saved objects client and generic API's in favour of plugins building custom HTTP API's. Previously auditing had to happen on the saved objects level, but if we're adding another layer into our architecture is that still the best place?

Maybe as a first step we can start by thinking through the ideal end state of audit logging and work back from that. This might be better suited to a sync discussion, but some questions we could explore:
 - If all data access happened through HTTP API's which either responded with 401, error or success could the URL and status code be an adequate replacement for generic saved_object_create/get/find/delete events?
 - there seems to be two classes of audit logs: generic saved_object_* events and events that are specific to the alerting domain, do we anticipate wanting to add more domain specific audit events?
 - from an auditing perspective what do we consider as interesting security events that needs to be logged. If a plugin has a hidden saved object exposed through an HTTP API and rejects a request because of some domain specific business rule, should this create an event? E.g. "only users with super-powers are allowed to break through steel doors"

Reply:
> Maybe as a first step we can start by thinking through the ideal end state of audit logging and work back from that. This might be better suited to a sync discussion, but some questions we could explore:

I forked this off into a shared doc to continue the discussion in a more sane medium than GH comment threads

Reply:
@rudolf I started describing some alternate approaches in [`682f6c8` (#93115)](https://github.com/elastic/kibana/pull/93115/commits/682f6c822926783e21786241fc9b2fb656b4bd6b)

Reply:
Of the alternatives, I'm currently favoring [5.2.2 - Pluggable authorization](https://github.com/legrego/kibana/blob/rfc/ols-phase-1/rfcs/text/0016_ols_phase_1.md#522-pluggable-authorization), but not necessarily coupled to this first phase of OLS. I think this model could work within the context of our "Persistence layer"/"Saved objects are data" discussions.

I'm not opposed to reevaluating if our timelines end up converging, though